### DISCUSSION TITLE
How to unit test the composition ?

### DISCUSSION THREAD

**Discussion by tlecomte** [QuestPDF/QuestPDF#689] [Q&A] [answered]
Following QuestPDF documentation, we have a data source, a model, and finally a document class that implements the `IDocument` interface.

We would like to implement some tests for the document class, since its implementation may be quite complex. For example, we would like to check that given some input model, the generated document will contain an expected string.

Is it possible to achieve this goal ?

With current QuestPDF implementation, it seems like it's not possible. The method to test would be `void IDocument.Compose(IDocumentContainer container);` . But the relevant elements in `IDocumentContainer` are all marked as internal, which prevents inspecting its content (see #665).

Alternatively, we thought about inspecting the generated pdf, but that's not practical. Similarly, inspecting the xps is difficult because it does not contain the raw text (see #688). 

Have you found other ways to achieve this ?

Thanks!

**Comment by MarcinZiabek** [comment, marked as answer]
It is a pretty interesting question. Currently, there is no official way of testing the document generation process. I recommended keeping all business logic outside of the generation logic, similar to the MVVM pattern. This way, the risk of potential mistakes in the QuestPDF-related code is minimized, while you can still easily unit test the business logic itself.

I am currently experimenting with [various testing approaches](https://github.com/QuestPDF/QuestPDF/tree/feat-layout-testing), but they are meant to be used internally during the initial phase. The idea is to compare commands performed on the canvas. So, instead of comparing bytes in the file stream or pixels on the image, I would like to verify the result at a higher level of abstraction.

That being said, there is an [official extension](https://www.nuget.org/packages/Verify.QuestPDF) for the Verify library that adds support for QuestPDF. It is certainly worth a try! Huge shout out to @SimonCropp for this fantastic work.

If you have any suggestions on how the testing suite should look for the library, please share your thoughts 😀

**Comment by DawidRobertNowak** [comment]
Hello,

We are closing this thread due to inactivity. Thank you for taking the time to reach out and contribute to the project — your feedback helps us improve QuestPDF.

Should you encounter any further problems, please don't hesitate to open a new issue.

Thank you!
